File: LeetCode/DataStructure/LRU,LFU/460.lfu-cache.py
#
# @lc app=leetcode id=460 lang=python3
#
# [460] LFU Cache
#

# @lc code=start
#
# @lc app=leetcode id=146 lang=python3
#
# [146] LRU Cache
#

# @lc code=start

class LHM_Node:
    def __init__(self, key, val, pNext, pPre) -> None:
        self.key = key
        self.val = val
        self.pNext = pNext
        self.pPre = pPre
        self.feq = 1 

# there are multiple of them, for each frequency
class LinkedList:
    def __init__(self) -> None:
        # container of LHM_Nodes grouped by their frequency
        self.pRecent = LHM_Node('recent', -1, None, None)
        self.pTail = LHM_Node('tail', -1, None, None)
        self.pRecent.pNext = self.pTail
        self.pTail.pPre = self.pRecent

class LinkedHashMap:
    def __init__(self, size) -> None:
        # keyToNode serve as container to all nodes, len(keyToNode) is current cache size
        # feqToLinkList serve as container to all LinkedLists, 

        self.size = size
        self.globalMinFeq = 1 # update when get/update(++cause prevLL empty) or insert (cause freq1LL non empty): gaurantee feqToLinkList[globalMinFeq] non empty
        self.feqToLinkList = {} # update when new freq appear (createNewLinkList())
        self.keyToNode = {} # update when del/insert node

    def insert(self, key, val):
        # delete
        if len(self.keyToNode) == self.size:
            # find least frequently used && least recently used
            if self.size == 0:
                return 
            minLL = self.feqToLinkList[self.globalMinFeq]
            delNode = minLL.pTail.pPre
            if (delNode is minLL.pRecent):
                raise ValueError("self.feqToLinkList[self.globalMinFeq] should not be empty")
            # move out of prevLL
            self.moveNodeOut(delNode)
            # possible require to update globalMinFeq here, tho, as globalMinFeq will soon be updated to 1, no worry
            # delete node from cache
            self.keyToNode.pop(delNode.key)


        # ! insert new node to LL freq == 1
        # create node and hashing
        newNode = LHM_Node(key, val, None, None)
        self.keyToNode[key] = newNode
        curLinkedList = None
        if 1 not in self.feqToLinkList:
            # create LinkedList and hashing
            curLinkedList = self.createNewLinkList(1)
        else:
            curLinkedList = self.feqToLinkList[1]
        # reset globalMinFeq
        self.globalMinFeq = 1
        # insert the node to head
        self.insertNodeToHead(newNode, curLinkedList)


    def update(self, key, val):
        curNode = self.keyToNode[key]
        self.frequentPlusOneOperations(curNode)
        curNode.val = val

    def put(self, key, val):
        if key in self.keyToNode:
            self.update(key, val)
        else:
            self.insert(key, val)

    def get(self, key: int) -> int:
        if key in self.keyToNode:
            curNode = self.keyToNode[key]
            self.frequentPlusOneOperations(curNode)
            return curNode.val
        else:
            return -1
           
    # ! utils
    def frequentPlusOneOperations(self, node):
        # ! update frq++, insert node to new LL and (possibly update global_min)
        # move out of prevLL
        self.moveNodeOut(node)

        # possible require to update globalMinFeq
        if node.feq == self.globalMinFeq:
            # check if prevLL is empty
            prevLL = self.feqToLinkList[node.feq] 
            if prevLL.pRecent.pNext is prevLL.pTail: # LL is empty
              
                self.globalMinFeq += 1 
        # insert curNode to new LinkedList
        node.feq += 1
        # may need to create new LinkedList 
        curLinkedList = None
        if node.feq not in self.feqToLinkList:
            curLinkedList = self.createNewLinkList(node.feq)
        else: 
            curLinkedList = self.feqToLinkList[node.feq]
        self.insertNodeToHead(node, curLinkedList)

        
    def createNewLinkList(self, freq):
        newLinkedList = LinkedList()
        self.feqToLinkList[freq] = newLinkedList
        return newLinkedList

    def insertNodeToHead(self, node, linkList):
        # insert the node to head
            # post node
        node.pNext = linkList.pRecent.pNext
        linkList.pRecent.pNext.pPre = node
            # pRecent(head)
        node.pPre = linkList.pRecent
        linkList.pRecent.pNext = node

    def moveNodeOut(self, node):
        node.pPre.pNext = node.pNext
        node.pNext.pPre = node.pPre


class LFUCache:

    def __init__(self, capacity: int):
        self.LHM = LinkedHashMap(size = capacity)
        
    def get(self, key: int) -> int:
        return self.LHM.get(key)

    def put(self, key: int, value: int) -> None:
        self.LHM.put(key, value)
        


# Your LFUCache object will be instantiated and called as such:
# obj = LFUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

# 数据按照访问频次进行排序
        
# 如果多个数据拥有相同的访问频次，删除最早插入的那个数据

# first inserted use counter is set to 1. The use counter is incremented either a get or put operation
    # each increment:
        # add freq by one
        # check if freqLinkList empty, and possibly update global minimal

File: 460.lfu-cache.java
/*
 * @lc app=leetcode id=460 lang=java
 *
 * [460] LFU Cache
 */

// @lc code=start

import java.util.HashMap;
import java.util.LinkedHashSet;

class Data {
    int frequency;
    int value;

    public Data(int value) {
        this.frequency = 0;
        this.value = value;
    }

    public int getFrequency() {
        return this.frequency;
    }

    public void setFrequency(int frequency) {
        this.frequency = frequency;
    }

    public int getValue() {
        return this.value;
    }

    public void setValue(int value) {
        this.value = value;
    }


}

class LFUCache {
    int minFreq;
    int capacity;
    HashMap<Integer, Data> keyToData;
    HashMap<Integer, LinkedHashSet<Integer>> freqToKeyRecencyArr;

    
    public LFUCache(int capacity) {
        this.minFreq = 0;
        this.capacity = capacity;
        this.keyToData = new HashMap<>();
        this.freqToKeyRecencyArr = new HashMap<>();
    }
    
    public int get(int key) {
        // return -1 if not exist
        // Data = keyToData[key]
        // addDataFreqByOne(key)

    }
    
    public void put(int key, int value) {
        // 1. update if the key already exist
            // 1.1 update
            // 1.2 add addDataFreqByOne
        // 2. insert if the key-value not already exist
            // check if cache reaches limit
            // 2.1 reaches limit, remove LF & LR Data
                // key = freqToKeyRecencyArr[minFreq].getTail
                // removeData(key)


            // 2.2 not reaches limit,
                // addNewData
            
    }

    // TODO
    private Void removeData(int key) {
        // remove key from keyToData
        // if 
    }

    private void addNewData(int key, int value) {
        // create new Data with frequency 1
        // insert to keyToData
        // insert the key into freqToKeyRecencyArr[1], create a new LinkedHashSet & add freqToKeyRecencyArr if needed
        // if minFreq not 1, update it to 1
    }
    private void addDataFreqByOne(int key) {
        // query the Data
        // update frequency in Data
        // move key of Data to the new LinkedHashSet(freq: +1), create new LinkedHashSet & add freqToKeyRecencyArr if needed
        // update minFreq (+1) if current minimal LinkedHashSet empty, free up the LinkedHashSet 
    }
}

/**
 * Your LFUCache object will be instantiated and called as such:
 * LFUCache obj = new LFUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
// @lc code=end

